### DISCUSSION TITLE
Configuring Dependabot with Reusable Workflows

### DISCUSSION THREAD

**Discussion by joshjohanning** [community/community#139004] [Discover: GitHub Best Practices]
## Overview

We can already use [Dependabot Version Updates](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/about-dependabot-version-updates) for keeping marketplace actions (in addition to internal/private actions) up to date. However, did you know we can use Dependabot for keeping Reusable Workflows up to date as well? 

## Authorization Options

If you are referring to reusable workflows that are public, [the implementation is easy](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot#example-dependabotyml-file-for-github-actions). More likely though, especially if you're in an organization creating reusable workflows to standardize on CI/CD practices, you are going to be working with non-public repositories containing your reusable workflows.

There are two ways that you can configure Dependabot when working with resources in internal or private repositories. To summarize, you can either:

1. Grant access via the UI
    - When Dependabot can't access a private repository, the logs allow you to [grant authorization to Dependabot to access your repository](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-security-and-analysis-settings-for-your-organization#allowing-dependabot-to-access-private-dependencies)
    - Alternatively, add it in the organization settings --> Code security and analysis --> [Grant Dependabot access to private repositories](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-security-and-analysis-settings-for-your-organization#allowing-dependabot-to-access-private-dependencies)
    - This setting requires organization admin permissions to access
    - Unfortunately, there isn't an API to automate this process, it has to  be done within the UI
2. Create a Dependabot secret
    - If you plan to create a large number repositories that you want to be a source for Dependabot, creating a [Dependabot secret](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/configuring-access-to-private-registries-for-dependabot#storing-credentials-for-dependabot-to-use) (preferably as an org-level Dependabot secret) with the value of a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) (preferably a [fine-grained token](https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/)) that has read-access to the required repositories would be preferred
    - I don't find it as detrimental to use a personal access token as a Dependabot secret since Dependabot secrets can *only be accessed by Dependabot*; you can't use a GitHub Actions workflow to expose the secret accidentally/intentionally
    - The only concern would be updating the token if it expires, is revoked, or the original author doesn't have access to the repo(s) anymore

> [!TIP]
> - If you only have a few, and rarely increasing set of repositories for custom actions / reusable workflows, I recommend the first approach. 
> - If you have a large number of repositories and/or are creating many new repositories for actions / reusable workflows, the **second approach scales better**.

### Approach 1: Grant Access to Dependabot in the UI

For the **first approach** (authorizing Dependabot access manually), the YML configuration is no different than if you were using Dependabot to keep [marketplace actions up to date](/posts/github-dependabot-for-actions/).

```yml
version: 2
updates:
  - package-ecosystem: "github-actions"
    directory: "/" # Workflow files stored in the default location of `.github/workflows`
    schedule:
      interval: "daily"
    open-pull-requests-limit: 5
```

You will then have to check your [Dependabot run logs](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/troubleshooting-dependabot-errors#investigating-errors-with-dependabot-version-updates) to authorize Dependabot for that repository (or [add it via the organization settings](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-security-and-analysis-settings-for-your-organization#allowing-dependabot-to-access-private-dependencies)):

![Noticing a Dependabot run failure](https://github.com/user-attachments/assets/dce8f060-37bc-4672-91be-e5d8c6b38a22)

![Granting access to Dependabot for this repository](https://github.com/user-attachments/assets/abecbc1a-0300-44eb-94c9-180bf093990c)

Once you grant Dependabot access to the repository, you will see it show up in organization settings –> Code security and analysis –> [Grant Dependabot access to private repositories](https://docs.github.com/en/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-security-and-analysis-settings-for-your-organization#allowing-dependabot-to-access-private-dependencies). Additional repositories can be added here:

![This repository now shows up under organization settings –> Code security and analysis –> Grant Dependabot access to private repositories](https://github.com/user-attachments/assets/23e84aab-f6a5-4d5d-8e2a-51bd0b5eba9d)

### Approach 2: Using a Dependabot Secret

For the **second option** (using a Dependabot secret), you will need to add the `registries` property to the YML configuration. The `registries` will reference `type: git` and use a [Dependabot Secret](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/configuring-access-to-private-registries-for-dependabot#storing-credentials-for-dependabot-to-use) (preferably an org-level Dependabot secret):

```yml
version: 2
updates:

  - package-ecosystem: "github-actions"
    directory: "/" # Workflow files stored in the default location of `.github/workflows`
    schedule:
      interval: "daily"
    registries:
      - github
registries:
  github:
    type: git
    url: https://github.com
    username: x-access-token # username doesn't matter
    password: ${{ secrets.MY_DEPENDABOT_SECRET }} # the dependabot secret you created
```

> [!NOTE]
> You can create a Dependabot secret at the organization level or repository level.

## Results

If things are working properly, you should see a successful run in your [Dependabot run logs](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/troubleshooting-dependabot-errors#investigating-errors-with-dependabot-version-updates):

![Example using Checks to create a line annotation](https://github.com/user-attachments/assets/9e512889-68a1-4c74-816c-26d7734a6a63)

And if there is a new semver version of a reusable workflow, you should see a Dependabot-created pull request:

![Example of a pull request for reusable workflow created by Dependabot](https://github.com/user-attachments/assets/f254b0a1-bb6d-4e89-951d-1a3b1baf0b38)

> [!TIP]
> Pro-tip: You can reply `@dependabot merge` or `@dependabot squash and merge` (among other commands) to tell Dependabot to merge the pull request.

## Summary

Now we can create and properly version reusable workflows AND have our downstream users automatically be notified of version updates. This helps a ton in making it front and center for developers that there's an update they need to look at! 🎉

**Comment by JorenVsyntra** [comment]
thanks mate!

**Comment by madhurigopal** [comment]
Check my repo on Real-Time Data Analytics. I think the issue we are facing here is the absence of a Versioning manager which Github already does. So,we may need to fine tune our SQL on our local and try using Dashboards to implement this as a feature or enhancement.
